joomla-projects/gsoc17_expand_extension_manager
# Download key manager

### Summary of Changes

Currently there is no clear overview of which extensions have download keys to make users download their extensions with an extra_query with their keys included to allow the download. Developers have been currently hacking this system by creating their own view to add a download key to the extra_query field in the #__update_sites table (Example of an extra_query field: dlid=%s&amp;dummy=my.zip (%s is the download key)).
Each extension developer may have a different way to build this extra query so the idea is to make a view to manage and insert the download keys of all extensions in a single view in com_installer.

### Testing Instructions

Here are the extensions to test the feature:

[Component](https://github.com/NunoLopes96/component_joomla)

Plugin:
[plg_csviaddon_virtuemart_csvi_7.1.0.zip](https://github.com/joomla-projects/gsoc17_expand_extension_manager/files/1177286/plg_csviaddon_virtuemart_csvi_7.1.0.zip)

Template:
[jpeople.zip](https://github.com/joomla-projects/gsoc17_expand_extension_manager/files/1177270/jpeople.zip)

Module:
[mod_weblinks.zip](https://github.com/joomla-projects/gsoc17_expand_extension_manager/files/1177265/mod_weblinks.zip)

Package:
[pkg_weblinks_3.4.1-changelog (3).zip](https://github.com/joomla-projects/gsoc17_expand_extension_manager/files/1177262/pkg_weblinks_3.4.1-changelog.3.zip)

Library:
[lib_fof30-3.1.0-changelog (1).zip](https://github.com/joomla-projects/gsoc17_expand_extension_manager/files/1177255/lib_fof30-3.1.0-changelog.1.zip)

File:
[cli2zoombie.zip](https://github.com/joomla-projects/gsoc17_expand_extension_manager/files/1177253/cli2zoombie.zip)


To allow the prefix and sufix of the download key you need to have an <dlid> tag with prefix and sufix in your installation file (here lists.xml) like this:

`<dlid prefix="dlid=" sufix="&amp;dummy=my.zip"/>`
(note: remember to close />)

Install the component and go to Extensions -> Download Keys Manager

![screenshot from 2017-07-16 17-57-50](https://user-images.githubusercontent.com/20955884/28249592-9c3121f4-6a50-11e7-9e23-65c2a4e64799.png)

You will not create or delete download keys so there will be only the edit button, the Download Key you see has the prefix and the suffix hidden, so the extra_query of this extension is:

`dlid=asd123asd&dummy=my.zip`

In the edit view, you will also have the prefix and sufix hidden, so the user don't have to worry about that. 
Only the developer will be responsible to define the prefix and sufix of the download key:

![screenshot from 2017-07-16 18-03-32](https://user-images.githubusercontent.com/20955884/28249631-4aef5a8a-6a51-11e7-8a3a-f9c48edb32ce.png)

### Modal button in com_plugins

The Downloadkeys view will probably be not known to all users, so to reinforce this new feature, a new modal button can be added to the extension when opened in com_plugins:

![screenshot from 2017-07-24 18-43-23](https://user-images.githubusercontent.com/20955884/28536601-08b185e4-70a0-11e7-9f92-4db4f6d77a6d.png)

Every plugin that has an update_site linked to it and a `<dlid>` tag on its xml installation file will pop this Download Key button in the toolbar, that will open a remote iframe to the DownloadKey view:

![screenshot from 2017-07-24 18-48-40](https://user-images.githubusercontent.com/20955884/28536817-bab648ba-70a0-11e7-989f-b746ab164782.png)

![screenshot from 2017-07-24 18-49-41](https://user-images.githubusercontent.com/20955884/28536870-de0188b6-70a0-11e7-8ef0-63c24f655c9a.png)

### Modal button in com_modules

The same will happen also to modules working the same way as com_plugins:
You can try this extension with an invalid updatesite and and dlid to testing porpuses:
[mod_weblinks.zip](https://github.com/joomla-projects/gsoc17_expand_extension_manager/files/1174805/mod_weblinks.zip)

![screenshot from 2017-07-25 22-27-52](https://user-images.githubusercontent.com/20955884/28594614-88eb1d04-7188-11e7-92e6-3b0d547e9374.png)
![screenshot from 2017-07-25 22-28-23](https://user-images.githubusercontent.com/20955884/28594633-96f50464-7188-11e7-976d-2f5540dfc5d0.png)
![screenshot from 2017-07-25 22-28-50](https://user-images.githubusercontent.com/20955884/28594656-a6965ada-7188-11e7-9b38-d14c32aba4d8.png)


### Expected result

A simple a intuitive way to the user manage his download keys and less work to the developers to create a view to generate the extra_query

### Actual result

### Documentation Changes Required
